File: src/Turis.BusinessLayer/Services/DocumentService.cs
using Microsoft.EntityFrameworkCore;
using Microsoft.EntityFrameworkCore.Query;
using Microsoft.Extensions.Logging;
using OperationResults;
using System.Globalization;
using JeMa.Shared.AspNetCore.Extensions;
using JeMa.Shared.Extensions;
using JeMa.Shared.Parameters.Base;
using TinyHelpers.Extensions;
using Turis.BusinessLayer.Extensions;
using Turis.BusinessLayer.Parameters;
using Turis.BusinessLayer.Services.Interfaces;
using Turis.Common.Enums;
using Turis.Common.Models;
using Turis.Common.Models.Requests;
using Turis.DataAccessLayer;
using Turis.DataAccessLayer.Entities;

namespace Turis.BusinessLayer.Services;

public class DocumentService(ApplicationDbContext dbContext
	, ILogger<DocumentService> logger
	, IUserService userService
	, IBookmarkService bookmarkService
	, IEntityTagService entityTagService
	) : IDocumentService
{
	private const string EntryName = nameof(Document);

	private readonly DbSet<Document> context = dbContext.Documents;

	private async Task<List<Bookmark>> GetMyBookmarks()
		=> await bookmarkService.ListAsync(userService.GetUserId(), EntryName);

	private IIncludableQueryable<Document, Contact> Query()
	{
		return context
			.Include(x => x.Items)
			.ThenInclude(x => x.Service)
			.Include(x => x.DocumentRef)
			.Include(x => x.Client)
			.Include(x => x.Collaborator);
	}

	public async Task<Result<DocumentModel>> GetAsync(Guid id)
	{
		var bookmarks = await bookmarkService.ListAsync(userService.GetUserId(), EntryName);

		var record = await Query()
			.FirstOrDefaultAsync(x => x.Id == id);

		if (record is null)
			return Result.Fail(FailureReasons.ItemNotFound);

		return await record.ToModelAsync(bookmarks);
	}

	public async Task<Result<DocumentModel>> GetAsync(string sectional, int number)
	{
		var bookmarks = await bookmarkService.ListAsync(userService.GetUserId(), EntryName);

		var record = await Query()
			.FirstOrDefaultAsync(x => x.Sectional == sectional && x.Number == number);

		if (record is null)
			return Result.Fail(FailureReasons.ItemNotFound);

		return await record.ToModelAsync(bookmarks);
	}

	public async Task<Result<PaginatedList<DocumentModel>>> ListAsync(DocumentSearchParameters parameters)
	{
		var bookmarks = await GetMyBookmarks();

		var paginator = new Paginator(parameters);

		var query = Query().AsNoTracking();

		if (parameters.OnlyBookmarks)
		{
			var bookmarkIds = bookmarks.Select(x => x.EntityId).ToList();
			query = query.WhereIf(parameters.OnlyBookmarks, x => bookmarkIds.Contains(x.Id));
		}

		if (parameters.DocumentType.HasValue())
		{
			var documentType = (DocumentType)Enum.Parse(typeof(DocumentType), parameters.DocumentType);
			query = query.Where(x => x.Type == documentType);
		}

		if (parameters.Sectional.HasValue())
		{
			query = query.Where(x => x.Sectional == parameters.Sectional);
		}

		if (parameters.NumberFrom > 0)
		{
			query = query.Where(x => x.Number >= parameters.NumberFrom);
		}
		if (parameters.NumberTo > 0)
		{
			query = query.Where(x => x.Number <= parameters.NumberTo);
		}

		if (parameters.DateFrom.HasValue())
		{
			var dateFrom = DateTime.ParseExact(parameters.DateFrom, "yyyyMMdd", CultureInfo.InvariantCulture);
			query = query.Where(x => x.Date >= dateFrom);
		}
		if (parameters.DateTo.HasValue())
		{
			var dateTo = DateTime.ParseExact(parameters.DateTo, "yyyyMMdd", CultureInfo.InvariantCulture);
			query = query.Where(x => x.Date <= dateTo);
		}

		// Filter by search pattern
		if (parameters.Pattern.HasValue())
			foreach (var itemPattern in parameters.Pattern.Split(' '))
			{
				var number = itemPattern.IsNumeric() ? Convert.ToInt32(itemPattern) : 0;

				query = query.Where(x =>
					(x.Sectional != null && x.Sectional.Contains(itemPattern))
					|| (number > 0 && x.Number == number)
					|| (number > 0 && x.Date.Year == number)
					|| (x.Client != null && x.Client.CompanyName.Contains(itemPattern))
				);
			}

		var totalCount = await query.CountAsync();

		if (parameters.OrderBy.HasValue())
		{
			//switch (parameters.OrderBy)
			//{
			//	case $"{nameof(Document.Name)}Asc":
			//		{
			//			query = query.OrderBy(x => x.Name);
			//			break;
			//		}
			//	case $"{nameof(Document.Name)}Desc":
			//		{
			//			query = query.OrderByDescending(x => x.Name);
			//			break;
			//		}
			//}
		}
		else
			query = query.OrderByDescending(x => x.Sectional)
				.ThenByDescending(x => x.Number);

		try
		{
			query = query.TakePage(parameters);

			var list = await query.ToListAsync();

			var model = await list.ToModelAsync(bookmarks);

			var result = model.ToPaginatedList(totalCount, parameters);

			return result;
		}
		catch (Exception ex)
		{
			Console.WriteLine(ex);
			throw;
		}
	}

	public async Task<Result> SaveAsync(DocumentRequest model)
	{
		var record = model.Id != Guid.Empty ? await context.FindAsync(model.Id) : null;
		if (record == null)
		{
			record = new Document
			{
				Id = Guid.NewGuid()
			};
			await context.AddAsync(record);
		}

		record.DocumentRefId = model.DocumentRefId;
		record.Type = (DocumentType)Enum.Parse(typeof(DocumentType), model.Type);
		record.Status = (DocumentStatus)Enum.Parse(typeof(DocumentStatus), model.Status);
		record.ClientId = model.ClientId;
		record.IdSdi = model.IdSdi;
		record.Date = model.Date;
		record.Sectional = model.Sectional;
		record.Number = model.Number;
		record.DiscountPercentage = model.DiscountPercentage;
		record.Discount = model.Discount;
		record.Amount = model.Amount;
		record.VatRate = model.VatRate;
		record.Vat = model.Vat;
		record.AliquotaRitenutaDiAcconto = model.AliquotaRitenutaDiAcconto;
		record.RitenutaDiAcconto = model.RitenutaDiAcconto;
		record.TotalExemptExpenses = model.TotalExemptExpenses;
		record.TotalExpenses = model.TotalExpenses;
		record.Total = model.Total;
		record.ImportoBollo = model.ImportoBollo;
		record.DesTipoPagamento = model.DesTipoPagamento;
		record.Saldato = model.Saldato;
		record.DataIncasso = model.DataIncasso;
		record.CollaboratorId = model.CollaboratorId;
		record.SdiCodiceTipoPagamento = model.SdiCodiceTipoPagamento;
		record.SdiValoreTipoPagamento = model.SdiValoreTipoPagamento;
		record.SdiCodiceCondizionePagamento = model.SdiCodiceCondizionePagamento;
		record.DataScadenzaPagamento = model.DataScadenzaPagamento;
		record.IdDocumento = model.IdDocumento;
		record.Cig = model.Cig;
		record.Cup = model.Cup;

		await entityTagService.UpdateTagsAsync(nameof(Service), record.Id, model.Tags);

		await dbContext.SaveChangesAsync();

		return Result.Ok();
	}

	public async Task<Result> DeleteAsync(Guid id)
	{
		await context.Where(x => x.Id == id).ExecuteDeleteAsync();
		return Result.Ok();
	}
}

File: src/Client/src/app/modules/document/document.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Document, DocumentSearchParameters } from '../document/document.types';
import { BaseEntityService } from 'app/shared/services';
import { BehaviorSubject, filter, finalize, map, Observable, of, switchMap, take, tap, throwError } from 'rxjs';
import { emptyGuid, PaginatedListResult } from 'app/shared/services/shared.types';

@Injectable({ providedIn: 'root' })
export class DocumentService extends BaseEntityService<Document> {
    private _edited: BehaviorSubject<string> = new BehaviorSubject(null);
    private _list: BehaviorSubject<PaginatedListResult<Document>> = new BehaviorSubject(null);
    private _item: BehaviorSubject<Document> = new BehaviorSubject(null);
    private _loading: BehaviorSubject<boolean> = new BehaviorSubject(false);
    private _parameters: BehaviorSubject<DocumentSearchParameters> = new BehaviorSubject({
        length: 0,
        pageIndex: 0,
        pageSize: 10,
    });

    constructor(http: HttpClient) {
        super(http);
        this.defaultApiController = 'document';
    }

    get list$(): Observable<PaginatedListResult<Document>> {
        return this._list.asObservable();
    }

    get item$(): Observable<Document> {
        return this._item.asObservable();
    }

    get loading$(): Observable<boolean> {
        return this._loading.asObservable();
    }

    get edited$(): Observable<string> {
        return this._edited.asObservable();
    }

    get parameters$(): Observable<DocumentSearchParameters> {
        return this._parameters.asObservable();
    }

    getById(id: string): Observable<Document> {
        return this.getSingle(id).pipe(
            map(document => {
                this._item.next(document);

                return document;
            }),
        );
    }

    createEntity(): Observable<Document> {
        const item: Document = {
            id: emptyGuid,
            documentRefId: undefined,
            documentRef: undefined,
            type: undefined,
            status: undefined,
            clientId: undefined,
            client: undefined,
            idSdi: undefined,
            date: undefined,
            sectional: undefined,
            number: undefined,
            discountPercentage: undefined,
            discount: undefined,
            amount: undefined,
            vatRate: undefined,
            vat: undefined,
            aliquotaRitenutaDiAcconto: undefined,
            ritenutaDiAcconto: undefined,
            totalExemptExpenses: undefined,
            totalExpenses: undefined,
            total: undefined,
            importoBollo: undefined,
            desTipoPagamento: undefined,
            saldato: undefined,
            dataIncasso: undefined,
            collaboratorId: undefined,
            collaborator: undefined,
            sdiCodiceTipoPagamento: undefined,
            sdiValoreTipoPagamento: undefined,
            sdiCodiceCondizionePagamento: undefined,
            dataScadenzaPagamento: undefined,
            idDocumento: undefined,
            cig: undefined,
            cup: undefined,
            bookmarkId: undefined,
            selected: false,
            items: [],
        };

        this._item.next(item);

        return of(item);
    }

    updateEntity(id: string, document: Document): Observable<Document> {
        return this.list$.pipe(
            take(1),
            switchMap(documents =>
                this.create(document).pipe(
                    map(() => {
                        // Find the index of the updated document
                        const index = documents.items.findIndex(item => item.id === id);

                        // Update the document
                        documents[index] = document;

                        // Update the document
                        this._item.next(document);

                        // Return the updated document
                        return document;
                    }),
                    switchMap(updatedDocument =>
                        this.item$.pipe(
                            take(1),
                            filter(item => item && item.id === id),
                            tap(() => {
                                // Update the document if it's selected
                                this._item.next(updatedDocument);

                                // Return the updated document
                                return updatedDocument;
                            }),
                        ),
                    ),
                ),
            ),
        );
    }

    listEntities(params?: DocumentSearchParameters): Observable<PaginatedListResult<Document>> {
        this._loading.next(true);

        let httpParams = new HttpParams();
        httpParams = httpParams.append('pageIndex', params?.pageIndex ?? 0);
        httpParams = httpParams.append('pageSize', params?.pageSize ?? 10);
        httpParams = httpParams.append('orderBy', params?.orderBy?.toString() ?? '');
        httpParams = httpParams.append('pattern', params?.pattern ?? '');
        httpParams = httpParams.append('onlyBookmarks', params?.onlyBookmarks ? 'true' : 'false');
        httpParams = httpParams.append('documentType', params?.documentType ?? '');
        httpParams = httpParams.append('sectional', params?.sectional ?? '');
        httpParams = httpParams.append('numberFrom', params?.numberFrom ?? 0);
        httpParams = httpParams.append('numberTo', params?.numberTo ?? 0);
        httpParams = httpParams.append('dateFrom', params?.dateFrom ?? '');
        httpParams = httpParams.append('dateTo', params?.dateTo ?? '');
        const queryString = httpParams.toString();

        const url = `?${queryString}`;

        return this.apiGet<PaginatedListResult<Document>>(url).pipe(
            map((list: PaginatedListResult<Document>) => {
                this._list.next(list);

                this._parameters.next({
                    ...this._parameters,
                    ...params,
                    pageIndex: list.pageIndex,
                    pageSize: list.pageSize,
                });

                return list;
            }),
            finalize(() => {
                this._loading.next(false);
            }),
        );
    }

    deleteEntity(id: string): Observable<Document> {
        return this._list.pipe(
            take(1),
            switchMap(documents => {
                // Remove the document
                this._item.next(null);

                // Remove the document from the _documents
                this._list.next({ ...documents, items: documents.items.filter(item => item.id !== id) });

                // Return the document
                return this.delete(id);
            }),
            switchMap(document => {
                if (!document) {
                    return throwError('Could not found document with id of ' + id + '!');
                }

                return of(document);
            }),
        );
    }

    editEntity(id: string): void {
        this._edited.next(id);
    }
}
